src/peergos/shared/io/ipfs/cid/Cid.java

```diff
@@ -49,35 +49,30 @@ public static Codec lookup(long c) {
     public final long version;
     public final Codec codec;
 
-    public Cid(long version, Codec codec, Multihash hash) {
-        super(hash.type, hash.getHash());
-        this.version = version;
-        this.codec = codec;
-    }
-
     public Cid(long version, Codec codec, Multihash.Type type, byte[] hash) {
         super(type, hash);
         this.version = version;
         this.codec = codec;
     }
 
-    public Cid(Multihash h) {
-        this(0, Codec.DagProtobuf, h);
+    public static Cid build(long version, Codec codec, Multihash h) {
+        return new Cid(version, codec, h.type, h.getHash());
     }
 
     private byte[] toBytesV0() {
         return super.toBytes();
     }
 
-    private static final int MAX_VARINT_LEN64 = 10;
-
     private byte[] toBytesV1() {
-        byte[] hashBytes = super.toBytes();
-        byte[] res = new byte[2 * MAX_VARINT_LEN64 + hashBytes.length];
-        int index = putUvarint(res, 0, version);
-        index = putUvarint(res, index, codec.type);
-        System.arraycopy(hashBytes, 0, res, index, hashBytes.length);
-        return Arrays.copyOfRange(res, 0, index + hashBytes.length);
+        try {
+            ByteArrayOutputStream res = new ByteArrayOutputStream();
+            putUvarint(res, version);
+            putUvarint(res, codec.type);
+            super.serialize(res);
+            return res.toByteArray();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
```

Comment:
Is runtime exception the best idea in this situation?

Reply:
It never throws an IOException because it is a byte array output stream.